### DISCUSSION TITLE
Can too many users on multiple devices cause issues or delays? Can mixing and matching client sites do the same?

### DISCUSSION THREAD

**Discussion by itemforty** [Lurkars/gloomhavensecretariat#325] [Q&A]
Our group is still having inconsistent, but substantial, delays when playing later in the evening on the weekends. Four times now it was so slow and buggy we've had to pivot and re-add an entire scenario to X-Haven, which is fine, but GHS is just the clearly better experience and we want it to work. However, I haven't seen anybody else in the community with this issue, so logically it has to be something we're doing. Here's the setup:

- 5 Android devices, different makes and ages. 
- 4 Players on 4 devices, 5th is using ChromeCast on a screen. 
- **Most** are using https://ghs.champonthis.de/ because X.ENV seems to be complete there
- Connecting to main public server
- On the monsters' turn, whoever is available will handle damage/conditions/etc.
- On our turn the active player will play while a different player (whoever is available) handles monster damage and loot cards. 
- Active player will input their XP, status effects, elements, and then tap their character face to indicate they are done with their turn.
- Repeat

Aside from extremely slow input delays, we've lost our undo history ("no state" IIRC), the last loot or modifier card drawn will get stuck in a loop, monster iconography disappears, dead figures return, HP changes are reverted, etc etc. Is this all caused by too many different people clicking too many different times? Clicking at the same time? Or something else? I would really love some help here because it's driving us crazy and we love GHS. 

Unrelated how difficult would it be to implement a non public server, WiFi only option, where one device is a host and the rest connect to it (like GHH and X-Haven have)? Just a thought. Thanks for anybody reading this!

**Comment by Lurkars** [comment]
I am just unsure if it's just the bad server connection to Germany. Maybe some ISPs perform different overseas that it doesn't affect everyone. Anyway, first have you tried to host your own server in local network? That would be the first thing come to my mind to check if it's only connection related or a different issue. Then always check that every client is on the same (latest) version. I don't test those cases, possible this also leads to issues.

**Reply by itemforty** [reply]
I haven't tried hosting on a local network yet. I assume that's following the instructions here https://github.com/Lurkars/ghs-server ? I'll try that out just to see. 

We always make sure we're on the latest version, but I didn't make sure everybody had the same URL, so I was wondering if that would matter. I noticed that on my Android **mobile** browser ENV-X isn't complete on gloomhaven-secretariat.de or gloomhaven-secretary.de (missing assets), but it seems to be fine on ghs.champonthis.de. Doesn't seem to matter on **desktop**.

It's also tough because when the app goes bad it's usually after multiple hours of play, and it becomes completely unusable. Hard to troubleshoot at that point. I wish I could figure out how to complete my private server, which eventually stopped letting us connect via the public SSL cert entirely (wouldn't let us bypass the security warning at all). I'm also not well enough versed to figure out how to take my Let's Encrypt keys in the application.properties file correctly or get the ports correct. If anybody has any step by step instructions on how to get that to work I would greatly appreciate it.

Anyway I'll try local next time we play at my house and go from there. Thanks for getting back to me!

**Reply by Lurkars** [reply]
Having different experience from ghs.champ to other domains can only be a weird cache issue, because technically they all point to the same Webserver with the same folder. I just kept old domains to not let anyone lose their state. So just reload a site if content is missing. Definitely impossible to have env-x data only on one site.

Therefore, different URL should also not be an issue. But env-x need to be enabled in every client separately, maybe you missed that.

Regarding SSL, if you use the [serve client](https://github.com/Lurkars/ghs-server#serve-the-client) functionality, you should also have a non-ssl client running and can access without SSL from there. So is you're hosting on local network, everyone can just use the local running client too then.

**Reply by itemforty** [reply]
Yeah I think you nailed it about enabling on the clients, that's smart. 

Using this thread for general info, if I were to use Bubblewrap to make an app out of the PWA, would the application still worry about a secure connection since Google has their own signed certs? If so do you know how configure it to work?

**Reply by Lurkars** [reply]
I cannot definetely say, but I think Bubblewrap will just create a webview opening the URL, so still https and still webview of your device and requirering ssl.
